spring-boot/proxy/src/test/java/security/ApiKeyAuthTokenVerifierTest.java

```diff
@@ -0,0 +1,189 @@
+    private ApiKey apiKey;
+
+    private LocalApiKeyManager localApiKeyManagerMock;
+    private ApiKeyFetcher apiKeyFetcherMock;
+    private ApiKeyVerifier apiKeyVerifierMock;
+
+    @BeforeEach
+    public void setup(){
+        apiKey = createSampleApiKeyForTesting();
+
+        apiKeyAuthTokenVerifier = createApiKeyAuthTokenVerifier();
+
+        // Create an auth token for a valid api key without any granted authorities.
+        apiKeyAuthToken = new ApiKeyAuthToken(someValidApiKeyValue, null);
+        apiKeyAuthToken.setAuthenticated(false);
+    }
+
+    private ApiKey createSampleApiKeyForTesting() {
+        ServiceApi serviceApi = new ServiceApi();
+        serviceApi.setUriIdentifier("weather-data-service");
+
+        ApiKey localApiKey = new ApiKey();
+        localApiKey.setKeyValue(someValidApiKeyValue);
+        localApiKey.setServiceApi(serviceApi);
+
+        return localApiKey;
+    }
+
+    private ApiKeyAuthTokenVerifier createApiKeyAuthTokenVerifier() {
+        localApiKeyManagerMock = mock(LocalApiKeyManager.class);
+        apiKeyFetcherMock = mock(ApiKeyFetcher.class);
+        apiKeyVerifierMock = mock(ApiKeyVerifier.class);
+
+        ApiKeyAuthTokenVerifier localApiKeyAuthTokenVerifier = new ApiKeyAuthTokenVerifier(
+                localApiKeyManagerMock,
+                apiKeyFetcherMock,
+                apiKeyVerifierMock
+        );
+
+        return localApiKeyAuthTokenVerifier;
+    }
+
+    @Test
+    public void nullArgumentShouldCauseException(){
+        assertThrows(SessionAuthenticationException.class, () -> apiKeyAuthTokenVerifier.authenticate(null));
+    }
+
+    @Test
+    public void nullCredentialsInAuthTokenShouldCauseException(){
+        ApiKeyAuthToken token = new ApiKeyAuthToken(null, null);
+
+        assertThrows(SessionAuthenticationException.class, () -> apiKeyAuthTokenVerifier.authenticate(token));
+    }
+
+    @Test
+    public void validApiKeyFromAdminShouldBeAuthenticated() throws ApiKeyFetchingFailedException {
+        extendApiKeyFetcherMock_SetExpectedResponse(ResultOfApiKeyRequestToAdmin.API_KEY);
+        extendApiKeyVerifierMock_ShallApiKeyBeValid(true);
+
+        assertTrue(apiKeyAuthTokenVerifier.authenticate(apiKeyAuthToken).isAuthenticated());
+    }
+
+    @Test
+    public void invalidApiKeyFromAdminShouldBeRejected() throws ApiKeyFetchingFailedException {
+        extendApiKeyFetcherMock_SetExpectedResponse(ResultOfApiKeyRequestToAdmin.API_KEY);
+        extendApiKeyVerifierMock_ShallApiKeyBeValid(false);
+
+        assertNull(apiKeyAuthTokenVerifier.authenticate(apiKeyAuthToken));
+    }
+
+    @Test
+    public void nullApiKeyReceivedFromAdminShouldBeRejected() throws ApiKeyFetchingFailedException {
+        extendApiKeyFetcherMock_SetExpectedResponse(ResultOfApiKeyRequestToAdmin.NULL);
+        extendApiKeyVerifierMock_ShallApiKeyBeValid(false);
+
+        assertNull(apiKeyAuthTokenVerifier.authenticate(apiKeyAuthToken));
```

Comment:
This should probably not return in a "null" but instead throw a BadCredentialsException?

Reply:
Yes. It is done now.